src/libraries/Common/src/Internal/Cryptography/PemKeyImportHelpers.cs

```diff
@@ -0,0 +1,157 @@
+            PemFields foundFields = default;
+            ReadOnlySpan<char> foundSlice = default;
+            bool containsEncryptedPem = false;
+
+            ReadOnlySpan<char> pem = input;
+            while (PemEncoding.TryFind(pem, out PemFields fields))
+            {
+                ReadOnlySpan<char> label = pem[fields.Label];
+                ImportKeyAction? action = callback(label);
+
+                // Caller knows how to handle this PEM by label.
+                if (action != null)
+                {
+                    // There was a previous PEM that could have been handled,
+                    // which means this is ambiguous and contains multiple
+                    // importable keys. Or, this contained an encrypted PEM.
+                    // For purposes of encrypted PKCS8 with another actionable
+                    // PEM, we will throw a duplicate exception.
+                    if (importAction != null || containsEncryptedPem)
+                    {
+                        throw new ArgumentException(SR.Argument_PemImport_AmbiguousPem, nameof(input));
+                    }
+
+                    importAction = action;
+                    foundFields = fields;
+                    foundSlice = pem;
+                }
+                else if (label.SequenceEqual("ENCRYPTED PRIVATE KEY"))
+                {
+                    containsEncryptedPem = true;
+                }
+
+                Index offset = fields.Location.End;
+                pem = pem[offset..];
+            }
+
+            // The only PEM found that could potentially be used is encrypted PKCS8,
+            // but we won't try to import it with a null or blank password, so
+            // throw.
+            if (containsEncryptedPem)
+            {
+                throw new ArgumentException(SR.Argument_PemImport_EncryptedPem, nameof(input));
+            }
+
+            // We went through the PEM and found nothing that could be handled.
+            if (importAction is null)
+            {
+                throw new ArgumentException(SR.Argument_PemImport_NoPemFound, nameof(input));
+            }
+
+            ReadOnlySpan<char> base64Contents = foundSlice[foundFields.Base64Data];
+            int base64size = foundFields.DecodedDataLength;
+            byte[] decodeBuffer = CryptoPool.Rent(base64size);
+
+            try
+            {
+                if (!Convert.TryFromBase64Chars(base64Contents, decodeBuffer, out int bytesWritten))
+                {
+                    // Couldn't decode base64. We shouldn't get here since the
+                    // contents are pre-validated.
+                    Debug.Fail("Base64 decoding failed on already validated contents.");
+                    throw new ArgumentException();
+                }
+
+                Debug.Assert(bytesWritten == base64size);
+                Span<byte> decodedBase64 = decodeBuffer.AsSpan(0, base64size);
+                importAction(decodedBase64, out _);
```

Comment:
I assume that you've made a conscious choice to say it's OK to ignore trailing bytes here, since there's no check that the import action reports bytesRead == base64Size.

I can accept that... but a comment saying so would be nice.  (Reasonable justification includes that the key has already been modified, so throwing and not have the same object modification effect)